### ISSUE TITLE
FR: Allow custom domains for Tailscale Funnel

### ISSUE THREAD

**Issue by weberc2** [tailscale/tailscale#11563]
### What are you trying to do?

I would like to expose a service to the public Internet using a custom domain name without requiring CNAME records, ideally via the Kubernetes operator.

### What is the impact of not solving this?

Workarounds require running a node with a public IP address to bridge to the tailnet and leaving users to manage their own certificates, which pretty much obviates the benefits of funnel as far as I can tell.

**Comment by ruifung** [comment]
If using kubernetes...

Just run a HAproxy that forwards traffic over tailscale using proxy protocol to your ingress controller of choice. (Letting the ingress controller handle TLS termination)

Pair it with cert-manager and external-dns and you got dont need to manually manage certificates and dns records. Just write Ingress objects or whatnot. Might need to annotate it with external dns annotations to properly assign the dns records but eh.

**Comment by renne** [comment]
It would be great to be able to use the organization domain and the subdomains for Tailscale Funnels.

**Comment by Kuvwert** [comment]
This feature is the only thing stopping me from moving from cloudflare tunnels to tailscale funnels.

If I could bring my own domains to TS Funnels I would be so happy

**Comment by Node815** [comment]
I am behind Cloudflare Tunnels and want to get away from them and Tailscale is the perfect solution for me, but in cases where I can't use it (locked down corporate PC with no admin rights), I can't install Tailscale to access my home network if I'm needing to troubleshoot something for the family.   With Cloudflare, I simply use my domain and I'm good to go.  Not so much with Tailscale.  

If custom domains were allowed , this would be ecstatic!

**Comment by mfechtner** [comment]
Could you not just use a cname on your domain to the tailscale domain name and serve a tcp port with your own ssl/tls termination.

Another option would be to spin up a small server yourself to to route traffic into tailscale.

**Comment by boomam** [comment]
You could maybe specify that the funnel could only be setup with Terraform,  forcing users to do a two part deploy that also deploys the DNS record to your nameserver provider perhaps?
Or have a list of compatible providers, and have TS auth via a token and have it setup the record?
Just throwing out ideas...

**Comment by mfechtner** [comment]
> You could maybe specify that the funnel could only be setup with Terraform,  forcing users to do a two part deploy that also deploys the DNS record to your nameserver provider perhaps?
> 
> Or have a list of compatible providers, and have TS auth via a token and have it setup the record?
> 
> Just throwing out ideas...

I assume the they have a rel. simple setup for the certificates based on sni and wildcards for their routing, which is probably more of a dynamic nature then a static deployment. 

Introducing custom domains may make this process more complex. Possibly limited by the capabilities of the underlying network infrastructure.

just guessing though...

**Comment by mfechtner** [comment]
> No, CNAME does not work because as stated above all tailscale sees (via SNI) is the CNAMEd name (mycooldomain.com) and does not know where to route the connection.
> SNI only indicates the requested domain, not the resolved alias.
> 
> So tailscale would need to support a way to "link" your custom domain name to your tailnet such that it can use it to look up the target endpoint when a SNI request comes in.
> 
> Or it would need to look up the dns name passed in via SNI itself, but that might have security implications I am not aware of.
> 
> Yes, spinning up your own VPS works but the funnel feature is specifically to avoid the need for this.

thanks, I rushed through the other comments. Will have a better look next time.

**Comment by Arusekk** [comment]
> Or it would need to look up the dns name passed in via SNI itself, but that might have security implications I am not aware of.

Incidentally, this is just how tunwg works:

https://github.com/ntnj/tunwg/blob/760ee8174aa95a6fa2ab93c192b8a1dca0a48bc2/tunwg/tunwgs.go#L186

**Comment by GerardUmbert** [comment]
I would deffinetly use this with a CNAME too. Many of us already have cloudflare, wouldn't be much of a hassle to have a cname and be able to avoid using tunnels. While we keep the logic in our infra regarding zero-trust rules and security measures without the need of a reverse proxy

**Comment by renne** [comment]
@GerardUmbert A solution for the domain apex would be great, too.

**Comment by Velocifyer** [comment]
> No, CNAME does not work because as stated above all tailscale sees (via SNI) is the CNAMEd name (mycooldomain.com) and does not know where to route the connection. SNI only indicates the requested domain, not the resolved alias.
> 
> So tailscale would need to support a way to "link" your custom domain name to your tailnet such that it can use it to look up the target endpoint when a SNI request comes in.
> 
> Or it would need to look up the dns name passed in via SNI itself, but that might have security implications I am not aware of.

You could just lookup the domain to see what it CNAME's to or use a txt record incase the users nameservers use cname flatening.

**Comment by Yiannis128** [comment]
Hi, has there been an update on this? It feels like a critical feature that should have priority.

**Comment by vuthanhtrung2010** [comment]
It's 2026, any updates on this?

**Comment by SmokeSlate** [comment]
I agree, any update, timeline?

**Comment by amithkk** [comment]
This would be amazing and would let me remove cloudflare tunnels from my infrastructure entirely

**Comment by binyaminyblatt** [comment]
Please implement this

**Comment by SmokeSlate** [comment]
Also, if possible being able to bind UDP/other TCP ports would be amazing. It would really help with hosting game and Web servers

**Comment by c3kuperus** [comment]
Is there any update on this as I am an individual Home Lab hobbyist that are looking to Set up my Home Server hosting to be "connected to the internet" like the funnel feature Proclaims to provide you but it seems like you cannot configure the TLS Termination to be Originating from a different domain then **"*.ts.net"**. It would be much cleaner and more helpful to expose services to the internet with your own domain than with the random configuration of Tailscale's MagicDNS. My current blocker is this specific configuration that is not allowed or not able to be configured on Tailscale side as Tailscale is terminating the connection.

**Comment by psyleft** [comment]
@c3kuperus I'm in the same situation and as far as I'm aware, this feature is required for that specific use case. In the mean time, what I have set up is to rent a cheap VPS and connect it to your tailnet, then use HAProxy or an equivalent that supports proxy protocol to route all TLS connections from ports 80 and 443 over the tailnet to a reverse proxy on the server side (I use traefik). It's still zero trust so you get to terminate SSL connections on your own machine and proxy protocol preserves request IPs, but of course it comes at the cost of paying for the VPS. I might publish my config files and do a write up on how to set everything up if it seems like lots of people are interested in doing the same.

**Comment by c3kuperus** [comment]
The way that I have overcome this issue is to use Tailscale Funnels for Protected Infrastructure (e.g. Personal computers and devices) and as I am already using CloudFlare as my DNS provider, I have opted to use CloudFlare Tunnel for the reverse proxy to host or self-host websites and web services. 

I have seen that it is mentioned to use a self-hosted reverse proxy, but inherently that brings some complexity to the setup and configuration. 

@psyleft, I do think that publishing any configuration files or documentation that could help anyone else that would want to dive into self-hosting reverse proxies would be very helpful to assist them. Thank you for your suggestion.

**Comment by c3kuperus** [comment]
Just as a last point that I would like to bring up is, I do believe that Tailscale is missing out on a wealth of Market share capture that for now is being redirected to other services. 

The reason why I originally opted to use Tailscale is because of the ease of use of setup and configuration.

**Comment by renne** [comment]
The other service is called Netbird reverse-proxy - including HTTP, HTTPS, TCP, UDP and TLS with custom domains in the most current beta release.

**Comment by truelockmc** [comment]
Couldn't this be implemented by allowing to expose the service URL (for example jellyfin.cool-fish.ts.net), instead of the device URL (for example raspberrypi.cool-fish.ts.net),

For example, a flag could be added to the tailscale-cli command, specifying if a service should be used as subdomain or something like that.

The main thing people would want to use funnels for is exposing things like mediaservers to the internet.
So wouldn*t it make more sense to just offer funneling the service URL instead of having to use the device URL?

I know this is not exactly a custom domain...
But it would be much less complicated to implement I think, and better than nothing.

**Comment by c3kuperus** [comment]
Yes, the service URL could definitely work and that's how I am using it currently for protected infrastructure but this thread is specifically related to custom domains. My issue was that setting up the Tailscale funnel does not allow a redirect from a custom domain, because of the SSL termination at Tailscale side, doing a validation on the domain origin and Tailscale does not allow you to configure a allowed origin list of some sorts.
